### PR TITLE
(FACT-1258) Guard all Ruby callbacks for exceptions

### PR DESCRIPTION
**Pull request by MikaelSmith**
Ruby doesn't know how to handle C++ exceptions (it's a C program) so we
must catch any C++ exceptions in the ruby_* calls to avoid crashing the
interpreter.

### CODE REVIEW

File: lib/src/ruby/module.cc [hunk 1]
```diff
@@ -444,357 +444,428 @@ namespace facter { namespace ruby {
...
+            }
+            return array;
+        });
     }
 
     VALUE module::ruby_which(VALUE self, VALUE binary)
     {
-        // Note: self is Facter::Core::Execution
-        auto const& ruby = api::instance();
+        return safe_eval("Facter.which", [&]() {
```
Comment (peterhuene) [inline lib/src/ruby/module.cc]: Technically this method is `Facter::Core::Execution::which`.

File: lib/src/ruby/module.cc [hunk 2]
```diff
@@ -444,357 +444,428 @@ namespace facter { namespace ruby {
...
+            return ruby.utf8_value(path);
+        });
     }
 
     VALUE module::ruby_exec(VALUE self, VALUE command)
     {
-        // Note: self is Facter::Core::Execution
-        auto const& ruby = api::instance();
-        return execute_command(ruby.to_string(command), ruby.nil_value(), false);
+        return safe_eval("Facter.exec", [&]() {
```
Comment (peterhuene) [inline lib/src/ruby/module.cc]: Technically this method is `Facter::Core::Execution::exec`.

File: lib/src/ruby/module.cc [hunk 3]
```diff
@@ -444,357 +444,428 @@ namespace facter { namespace ruby {
...
+            auto const& ruby = api::instance();
+            return execute_command(ruby.to_string(command), ruby.nil_value(), false);
+        });
     }
 
     VALUE module::ruby_execute(int argc, VALUE* argv, VALUE self)
     {
-        // Note: self is Facter::Core::Execution
-        auto const& ruby = api::instance();
+        return safe_eval("Facter.exec", [&]() {
```
Comment (peterhuene) [inline lib/src/ruby/module.cc]: Technically this method is `Facter::Core::Execution::execute`.

**Comment by MikaelSmith** [comment]
Hah, should've told you to ignore whitespace. I can fix those up though.

**Comment by MikaelSmith** [comment]
Updated.

**Pull request merged**